sql: new EXPLAIN infrastructure

RaduBerinde opened the pull request:
#### opt: store columns in OpaqueMetadata

Store the result columns of opaque operators in their metadata instead
of returning it separately.

Release note: None

#### opt: separate CreateTable/CreateTableAs exec operators

Separate the two variants in the exec factory so that we don't have to
pass around a `nil` node.

Release note: None

#### sql: new EXPLAIN infrastructure

This change reimplements the default `(PLAN)` variant of EXPLAIN.

Instead of visiting a plan node tree to get EXPLAIN information, we
record the exact `exec.Factory` calls that were produced. We do this
using an `exec.ExplainFactory` which is an extension of
`exec.Factory`.

This factory wraps a regular `exec.Factory` and forwards all factory
calls to it (while also recording all arguments). The result is that
for each `exec.Node` there is a corresponding `*explain.Node`. The
factory also provides a way for the execbuilder to annotate nodes with
more information, which will be initially used for row count and cost
(not exposed in this PR).

The implementation relies on optgen for the factory methods and
per-operation identifiers and structs.

I tried to avoid major changes to the EXPLAIN output; I left a lot of
TODOs for improvements that I plan to make in follow-up changes. Some
egregious issues were fixed in separate PRs. Still, there are various
differences; the notable ones are:
 - for simple projections, we now show a `project` node without any
   rendering expressions.
 - in some cases we see a new `project` that just rearranges columns.
   I will address this in a follow-up change.
 - window function information is no longer duplicated.
 - expressions for inverted filters and joins now show the column
   names instead of unnamed references like `@2`.
 - we no longer see a `count` node on top of mutations.

Release note (sql change): various improvements to EXPLAIN.

#### sql: EXPLAIN test updates

Separate commit for explain test changes.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/52730)
<!-- Reviewable:end -->

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> very nice work!

> Reviewed 3 of 3 files at r1, 4 of 4 files at r2, 21 of 21 files at r3, 41 of 41 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-:-ME_2p003tg7k1-hpqMA:b-5u6ws8)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @jordanlewis, @mgartner, and @RaduBerinde)

___
*[pkg/sql/opt_exec_factory.go, line 1662 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZrJbG52dkCbvyibgg:-MEZrJbG52dkCbvyibgh:b2u99ks) ([raw file](https://github.com/cockroachdb/cockroach/blob/04bac038813c05f7d53618a46d5ac659d5de4a05/pkg/sql/opt_exec_factory.go#L1662)):*
> ```Go
> }
> 
> //
> ```

nit: empty comment.
___
*[pkg/sql/opt/exec/factory.opt, line 535 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZrOFW5yBSz2N5hTaW:-MEZrOFW5yBSz2N5hTaX:br2gsa3) ([raw file](https://github.com/cockroachdb/cockroach/blob/04bac038813c05f7d53618a46d5ac659d5de4a05/pkg/sql/opt/exec/factory.opt#L535)):*
> ```
> }
> 
> # CreateTable implements a CREATE TABLE AS statement.
> ```

nit: `s/CreateTable/CreateTableAs/g`.
___
*[pkg/sql/opt/exec/explain/emit.go, line 27 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZtXev6CgKFaPW_mij:-MEZtXev6CgKFaPW_mik:b1hdiw6) ([raw file](https://github.com/cockroachdb/cockroach/blob/03710f5fdd61658ab0efb1fa440fc741ed2b870a/pkg/sql/opt/exec/explain/emit.go#L27)):*
> ```Go
> )
> 
> // Emit produces the EXPLAIN output against the given OutputBuilder. The flags
> ```

nit: something is off in "flags OutputBuilder flags".
___
*[pkg/sql/opt/exec/explain/emit.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZth3D35T-UOxn56uq:-MEZth3D35T-UOxn56ur:bn4ylmm) ([raw file](https://github.com/cockroachdb/cockroach/blob/03710f5fdd61658ab0efb1fa440fc741ed2b870a/pkg/sql/opt/exec/explain/emit.go#L31)):*
> ```Go
> func Emit(plan *Plan, ob *OutputBuilder, spanFormatFn SpanFormatFn) error {
> 	e := makeEmitter(ob, spanFormatFn)
> 	var walk func(n *Node) error
> ```

super nit: is there a reason to have a separate variable declaration?
___
*[pkg/sql/opt/exec/explain/emit.go, line 311 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZw4QR5vWaGxpMmjol:-MEZw4QR5vWaGxpMmjom:b-5f27ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/03710f5fdd61658ab0efb1fa440fc741ed2b870a/pkg/sql/opt/exec/explain/emit.go#L311)):*
> ```Go
> 		ob.Attr("order", sqlbase.ColumnOrdering(a.Ordering).String(n.Columns()))
> 		if p := a.AlreadyOrderedPrefix; p > 0 {
> 			ob.Attr("already ordered", sqlbase.ColumnOrdering(a.Ordering[p:]).String(n.Columns()))
> ```

Shouldn't this be `a.Ordering[:p]`?
___
*[pkg/sql/opt/exec/explain/explain_factory.go, line 102 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZyQdy0DGXvoDV4R3w:-MEZyQdy0DGXvoDV4R3x:b-dyfb79) ([raw file](https://github.com/cockroachdb/cockroach/blob/03710f5fdd61658ab0efb1fa440fc741ed2b870a/pkg/sql/opt/exec/explain/explain_factory.go#L102)):*
> ```Go
> 
> // Plan is the result of ConstructPlan; provides access to the exec.Plan
> // prodChildCountuced by the wrapped factory.
> ```

nit: "prodChildCountuced".
___
*[pkg/sql/opt/exec/explain/result_columns.go, line 24 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZzp2U5oVYTxuPn1e1:-MEZzp2U5oVYTxuPn1e2:brhuabw) ([raw file](https://github.com/cockroachdb/cockroach/blob/03710f5fdd61658ab0efb1fa440fc741ed2b870a/pkg/sql/opt/exec/explain/result_columns.go#L24)):*
> ```Go
> )
> 
> //
> ```

nit: empty comment line.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: store columns in OpaqueMetadata

Store the result columns of opaque operators in their metadata instead
of returning it separately.

Release note: None

RaduBerinde committed:
opt: separate CreateTable/CreateTableAs exec operators

Separate the two variants in the exec factory so that we don't have to
pass around a `nil` node.

Release note: None

RaduBerinde committed:
sql: new EXPLAIN infrastructure

This change reimplements the default `(PLAN)` variant of EXPLAIN.

Instead of visiting a plan node tree to get EXPLAIN information, we
record the exact `exec.Factory` calls that were produced. We do this
using an `exec.ExplainFactory` which is an extension of
`exec.Factory`.

This factory wraps a regular `exec.Factory` and forwards all factory
calls to it (while also recording all arguments). The result is that
for each `exec.Node` there is a corresponding `*explain.Node`. The
factory also provides a way for the execbuilder to annotate nodes with
more information, which will be initially used for row count and cost
(not exposed in this PR).

The implementation relies on optgen for the factory methods and
per-operation identifiers and structs.

I tried to avoid major changes to the EXPLAIN output; I left a lot of
TODOs for improvements that I plan to make in follow-up changes. Some
egregious issues were fixed in separate PRs. Still, there are various
differences; the notable ones are:
 - for simple projections, we now show a `project` node without any
   rendering expressions.
 - in some cases we see a new `project` that just rearranges columns.
   I will address this in a follow-up change.
 - window function information is no longer duplicated.
 - expressions for inverted filters and joins now show the column
   names instead of unnamed references like `@2`.
 - we no longer see a `count` node on top of mutations.

Release note (sql change): various improvements to EXPLAIN.

RaduBerinde committed:
sql: EXPLAIN test updates

Separate commit for explain test changes.

Release note: None

RaduBerinde reviewed (commented):
TFTR!

Fixed a few things that were causing test failures (e.g. EXPLAIN inside EXPLAIN).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-:-ME_EYWLDNGV09WmhBn8:bhz991q)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @jordanlewis, @mgartner, and @yuzefovich)

___
*[pkg/sql/opt/exec/explain/emit.go, line 31 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZth3D35T-UOxn56uq:-ME_5amVD44uU2jfHOQf:bnh88fs) ([raw file](https://github.com/cockroachdb/cockroach/blob/03710f5fdd61658ab0efb1fa440fc741ed2b870a/pkg/sql/opt/exec/explain/emit.go#L31)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: is there a reason to have a separate variable declaration?
</blockquote></details>

it's necessary because it's recursive
___
*[pkg/sql/opt/exec/explain/emit.go, line 311 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/52730#-MEZw4QR5vWaGxpMmjol:-ME_5mvi9Thxp6lOXKOb:b-wyl0xv) ([raw file](https://github.com/cockroachdb/cockroach/blob/03710f5fdd61658ab0efb1fa440fc741ed2b870a/pkg/sql/opt/exec/explain/emit.go#L311)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Shouldn't this be `a.Ordering[:p]`?
</blockquote></details>

Wow nice catch, especially given that it was buried in all this tedious new code :)


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
